### ISSUE TITLE
FR: Consolidate the numerous sd_setImage signatures

### ISSUE THREAD

**Issue by willbattel** [firebase/FirebaseUI-iOS#919]
### Step 1: Are you in the right place?

Yes, but an issue in the SDWebImage repository may also be warranted- whether opened by me or a FirebaseUI contributor. I didn't want to jump the gun before getting feedback here.

### Step 2: Describe your environment

  * Objective C or Swift: Swift 5.3
  * iOS version: 11.2
  * Firebase SDK version: 7.1.0
  * FirebaseUI version: 9.0.0
  * CocoaPods Version: 1.10.0
  
### Step 3: Describe the problem:

FirebaseUI integrates with SDWebImage's `sd_setImage` function to serve images from Firebase Storage directly into any UIImageView. This is a really nifty feature, and my complaint here is more of a minor grievance than anything else.

When exploiting the `sd_setImage` function, the developer will likely notice that there are many signatures for this function; I counted 12 in total, 7 of which use Firebase's `StorageReference` and the other 5 being SDWebImage's native URL-based variants. These signatures vary in parameters such as `placeholderImage` and `options`, but in doing so create a bit of a mess. There are two parts to this mess:

1. Finding the best function signature to use for my specific situation takes longer than desired; and
2. Some situations will require a signature that provides additional parameters that I don't want to use.

<img width="657" alt="Screen Shot 2020-11-23 at 1 49 40 AM" src="https://user-images.githubusercontent.com/13490339/99935413-a54d8a80-2d2e-11eb-97d0-de18e94950bd.png">

For example, I have a UIImageView where I want to download from a `StorageReference`, enable the `SDWebImageRetryFailed` option, and provide a completion closure. Unfortunately, there is no signature with only these parameters. The best option currently is to use `sd_setImage(with:maxImageSize:placeholderImage:options:completion:)`. This requires me to specify `maxImageSize` and `placeholderImage` even though I have no interest in setting either. Fortunately in this case, I can simply use `.max` and `nil`, respectively, to effectively ignore these parameters- but I stand by my belief that I should not have to specify them in the first place.

My proposal is to have only 2 signatures instead of the current 12, one for URL-based requests and one for StorageReference-based requests. Then, each of these would provide all current parameters but with default values to make them ignorable at the call site.

Let's look at the existing signature that includes all available parameters:

`sd_setImage(with: StorageReference, maxImageSize: UInt64, placeholderImage: UIImage?, options: SDWebImageOptions, context: [SDWebImageContextOption : Any]?, progress: ((Int, Int, StorageReference) -> Void)?, completion: ((UIImage?, Error?, SDImageCacheType, StorageReference) -> Void)?)`

It seems to me that each of these parameters, except for the source, have a reasonable default value that could be used:

`maxImageSize = .max`
`placeholderImage = nil`
`options = []`
`context = nil`
`progress = nil`
`completion = nil`

Using default values under a single function would relieve both of the two pain points listed above. I'd love to get some feedback, maybe there's a larger reason for the overloads that I'm missing.

<hr>

**tl;dr:** There are too many `sd_setImage` overloads that make using it unnecessarily annoying. Being only a minor grievance, I simply request that at some point the FirebaseUI and/or SDWebImage teams will consider reorganizing them.

**Comment by morganchen12** [comment]
Given our constraint of writing this library in Objective-C, I'm not sure what you're suggesting is possible. Specifically, the `Using default values under a single function` feature isn't available in ObjC, and I think in Swift it will generate separate auto-completable functions anyway.

Also, as you've already mentioned, 5 of the 12 function overloads are provided by SDWebImage itself, and the remaining 7 come from FirebaseUI. For a more pleasant autocomplete experience, you can open up [the header file](https://github.com/firebase/FirebaseUI-iOS/blob/master/Storage/FirebaseStorageUI/UIImageView%2BFirebaseStorage.h) and comment out all but the function with the most parameters if that's the only function you care about.

My current understanding of your request is it's infeasible; let me know if I'm missing anything.

**Comment by willbattel** [comment]
> My current understanding of your request is it's infeasible; let me know if I'm missing anything.

No, I think you're correct. I just wanted to document this as a pain point so that if there were ever a time in the future where this becomes feasible then it would already be on radar as something someone had complained about. 

Anything that makes me stop and think to myself "_that's annoying_" usually leads me to open an issue, for posterity's sake if nothing else 🙃

**Comment by morganchen12** [comment]
No problem, thanks for being proactive in leaving feedback and myself and the other Firebase staff appreciate your thoroughly-written issues.
